fem/bilinearform_ext.hpp

```diff
@@ -39,99 +39,108 @@ class ElemRestriction: public Operator
 {
-private:
-   BilinearForm *a;
 public:
-   EABilinearFormExtension(BilinearForm *form);
+   EABilinearFormExtension(BilinearForm *form)
+      : BilinearFormExtension(form) { }
 
    /// TODO
-   void AddDomainIntegrator(BilinearFormIntegrator*) {}
    void Assemble() {}
-   void FormSystemOperator(const Array<int> &ess_tdof_list, Operator *&A) {}
+   void FormSystemMatrix(const Array<int> &ess_tdof_list, OperatorHandle &A) {}
    void FormLinearSystem(const Array<int> &ess_tdof_list,
                          Vector &x, Vector &b,
-                         Operator *&A, Vector &X, Vector &B,
+                         OperatorHandle &A, Vector &X, Vector &B,
                          int copy_interior = 0) {}
-   void RecoverFEMSolution(const Vector &X, const Vector &b, Vector &x) {}
    void Mult(const Vector &x, Vector &y) const {}
    void MultTranspose(const Vector &x, Vector &y) const {}
    ~EABilinearFormExtension() {}
 };
 
 /// Data and methods for partially-assembled bilinear forms
-class PABilinearFormExtension : public Operator
+class PABilinearFormExtension : public BilinearFormExtension
 {
-private:
-   BilinearForm *a;
+protected:
    const FiniteElementSpace *trialFes, *testFes;
-   Array<BilinearFormIntegrator*> integrators;
    mutable Vector localX, localY;
    ElemRestriction *elem_restrict;
 
 public:
    PABilinearFormExtension(BilinearForm*);
-   void AddDomainIntegrator(BilinearFormIntegrator*);
-   // void AddBoundaryIntegrator(BilinearFormIntegrator*);
-   // void AddInteriorFaceIntegrator(BilinearFormIntegrator*);
-   // void AddBoundaryFaceIntegrator(BilinearFormIntegrator*);
 
    void Assemble();
-   void FormSystemOperator(const Array<int> &ess_tdof_list, Operator *&A);
+   void FormSystemMatrix(const Array<int> &ess_tdof_list, OperatorHandle &A);
    void FormLinearSystem(const Array<int> &ess_tdof_list,
                          Vector &x, Vector &b,
-                         Operator *&A, Vector &X, Vector &B,
+                         OperatorHandle &A, Vector &X, Vector &B,
                          int copy_interior = 0);
-   void RecoverFEMSolution(const Vector &X, const Vector &b, Vector &x);
 
    void Mult(const Vector &x, Vector &y) const;
    void MultTranspose(const Vector &x, Vector &y) const;
-   void Update(FiniteElementSpace *fes);
+   void Update();
 
    ~PABilinearFormExtension();
 };
 
 /// Data and methods for matrix-free bilinear forms
-class MFBilinearFormExtension : public Operator
+class MFBilinearFormExtension : public BilinearFormExtension
 {
-private:
-   BilinearForm *a;
 public:
-   MFBilinearFormExtension(BilinearForm *form);
+   MFBilinearFormExtension(BilinearForm *form)
+      : BilinearFormExtension(form) { }
 
    /// TODO
-   void AddDomainIntegrator(BilinearFormIntegrator*) {}
    void Assemble() {}
-   void FormSystemOperator(const Array<int> &ess_tdof_list, Operator *&A) {}
+   void FormSystemMatrix(const Array<int> &ess_tdof_list, OperatorHandle &A) {}
```

Comment:
`FormSystemMatrix` is quite a misnomer in this case, which is why we thought `FormSystemOperator` may be better... (I don't insist on changing back the name though, I can live with `FormSystemMatrix`)